AsyncDisplayKit/Layout/ASStackLayoutNode.mm

```diff
@@ -0,0 +1,133 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "ASStackLayoutNode.h"
+
+#import <numeric>
+#import <vector>
+
+#import "ASBaseDefines.h"
+#import "ASInternalHelpers.h"
+
+#import "ASLayoutNodeUtilities.h"
+#import "ASLayoutNodeSubclass.h"
+#import "ASStackLayoutNodeUtilities.h"
+#import "ASStackPositionedLayout.h"
+#import "ASStackUnpositionedLayout.h"
+
+@implementation ASMutableStackLayoutNodeChild
+@synthesize node, spacingBefore, spacingAfter, flexGrow, flexShrink, flexBasis, alignSelf;
+@end
+
+@implementation ASStackLayoutNodeChild
+
+- (instancetype)initWithNode:(ASLayoutNode *)node
+               spacingBefore:(CGFloat)spacingBefore
+                spacingAfter:(CGFloat)spacingAfter
+                    flexGrow:(BOOL)flexGrow
+                  flexShrink:(BOOL)flexShrink
+                   flexBasis:(ASRelativeDimension)flexBasis
+                   alignSelf:(ASStackLayoutAlignSelf)alignSelf
+{
+  if (node == nil)
+    return nil;
+  
+  if (self = [super init]) {
+    _node = node;
+    _spacingBefore = spacingBefore;
+    _spacingAfter = spacingAfter;
+    _flexGrow = flexGrow;
+    _flexShrink = flexShrink;
+    _flexBasis = flexBasis;
+    _alignSelf = alignSelf;
+  }
+  return self;
+}
+
+- (id)copyWithZone:(NSZone *)zone
+{
+  if ([self isKindOfClass:[ASMutableStackLayoutNodeChild class]]) {
+    return [[ASStackLayoutNodeChild alloc] initWithNode:self.node
+                                          spacingBefore:self.spacingBefore
+                                           spacingAfter:self.spacingAfter
+                                               flexGrow:self.flexGrow
+                                             flexShrink:self.flexShrink
+                                              flexBasis:self.flexBasis
+                                              alignSelf:self.alignSelf];
+  } else {
+    return self;
+  }
+}
+
+- (id)mutableCopyWithZone:(NSZone *)zone
+{
+  ASMutableStackLayoutNodeChild *mutableChild = [[ASMutableStackLayoutNodeChild alloc] init];
+  mutableChild.node = self.node;
+  mutableChild.spacingBefore = self.spacingBefore;
+  mutableChild.spacingAfter = self.spacingAfter;
+  mutableChild.flexGrow = self.flexGrow;
+  mutableChild.flexShrink = self.flexShrink;
+  mutableChild.flexBasis = self.flexBasis;
+  mutableChild.alignSelf = self.alignSelf;
+  return mutableChild;
+}
+
++ (instancetype)newWithInitializer:(void (^)(ASMutableStackLayoutNodeChild *))initializer
+{
+  ASStackLayoutNodeChild *c = [super new];
```

Comment:
Sure. This builder pattern can be used to create a big data object (ASStackLayoutNodeChild in this case) with most of its properties have default value. It also allows users to create the object in an inline fashion, as opposed to create using a default initializer and then change a few needed properties. 

This pattern is the result of borrowed ideas from [this blog post](http://cocoanuts.mobi/2014/05/15/builder/) and [Pop](https://github.com/facebook/pop/blob/master/pop/POPAnimatableProperty.mm#L1244).